core/renderers/measurables/types.js

```diff
@@ -0,0 +1,353 @@
+ *
+ * Copyright 2019 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Measurable types.
+ * @author samelh@google.com (Sam El-Husseini)
+ */
+
+'use strict';
+
+goog.provide('Blockly.blockRendering.Types');
+
+
+/**
+ * Types of rendering elements.
+ * @enum {number}
+ * @package
+ */
+Blockly.blockRendering.Types = {
+  NONE: 0, // None
+  FIELD: 1 << 0, // Field.
+  HAT: 1 << 1, // Hat.
+  ICON: 1 << 2, // Icon.
+  SPACER: 1 << 3, // Spacer.
+  BETWEEN_ROW_SPACER: 1 << 4, // Between Row Spacer.
+  IN_ROW_SPACER: 1 << 5, // In Row Spacer.
+  EXTERNAL_VALUE_INPUT: 1 << 6, // External Value Input.
+  INPUT: 1 << 7, // Input
+  INLINE_INPUT: 1 << 8, // Inline Input.
+  STATEMENT_INPUT: 1 << 9, // Statement Input.
+  CONNECTION: 1 << 10, // Connection.
+  PREVIOUS_CONNECTION: 1 << 11, // Previous Connection.
+  NEXT_CONNECTION: 1 << 12, // Next Connection.
+  OUTPUT_CONNECTION: 1 << 13, // Output Connection.
+  CORNER: 1 << 14, // Corner.
+  LEFT_SQUARE_CORNER: 1 << 15, // Square Corner.
+  LEFT_ROUND_CORNER: 1 << 16, // Round Corner.
+  RIGHT_SQUARE_CORNER: 1 << 17, // Right Square Corner.
+  RIGHT_ROUND_CORNER: 1 << 18, // Right Round Corner.
+  JAGGED_EDGE: 1 << 19, // Jagged Edge.
+  ROW: 1 << 20, // Row
+  TOP_ROW: 1 << 21, // Top Row.
+  BOTTOM_ROW: 1 << 22, // Bototm Row.
+  INPUT_ROW: 1 << 23, // Input Row.
+};
+
+/**
+ * A Left Corner Union Type.
+ * @type {number}
+ * @const
+ * @package
+ */
+Blockly.blockRendering.Types.LEFT_CORNER =
+    Blockly.blockRendering.Types.LEFT_SQUARE_CORNER |
+    Blockly.blockRendering.Types.LEFT_ROUND_CORNER;
+
+/**
+ * A Right Corner Union Type.
+ * @type {number}
+ * @const
+ * @package
+ */
+Blockly.blockRendering.Types.RIGHT_CORNER =
+    Blockly.blockRendering.Types.RIGHT_SQUARE_CORNER |
+    Blockly.blockRendering.Types.RIGHT_ROUND_CORNER;
+
+/**
+ * Next flag value to use for custom rendering element types.
+ * This must be updated to reflect the next enum flag value
+ * to use if additional elements are added to
+ * `Blockly.blockRendering.Types`.
+ * @type {number}
+ * @private
+ */
+Blockly.blockRendering.Types.nextTypeValue_ = 1 << 24;
+
+/**
+ * Get the enum flag value of an existing type or register a new type.
+ * @param {!string} type The name of the type.
+ * @return {!number} The enum flag value assosiated with that type.
+ * @package
+ */
+Blockly.blockRendering.Types.getType = function(type) {
```

Comment:
When is this used?

Reply:
Also, why is the name a string here?  Can you index into an enum with a string?

Reply:
An enum is just an object so yes. and it's not used anywhere right now but it's for custom (dynamic) rendering types. I already have a few cases for this, but in essence, this avoids polluting these types with element types that are specific to some renderers.
Instead they can use create / register a type by doing something like: 
``
this.type = Blockly.blockRendering.Types.getType('CUSTOM_ELEMENT_TYPE');
```
and then to check in their code if an element is of that type: 
```
elem.type & Blockly.blockRendering.Types.getType('CUSTOM_ELEMENT_TYPE')
```

Reply:
SGTM